src/cli/repl/meta_cmd.rs

```diff
@@ -649,8 +669,503 @@ impl<F: LurkField + DeserializeOwned> MetaCmd<F> {
+        let Some((args_vec, None)) = repl.store.fetch_list(args) else {
+            bail!("Protocol arguments must be a proper list")
+        };
+
+        if args_vec.len() != len {
+            bail!("Incompatible number of arguments. Protocol expects {len}")
+        }
+        Ok(args_vec)
+    }
+
+    /// Runs the protocol `body` with an environment built from the bindings
+    /// creted with `vars` and `args` and returns the proof input and the
+    /// post-verify function.
+    ///
+    /// # Errors
+    /// * If the protocol body evaluation fails
+    /// * If the protocol body doesn't evaluate to a pair
+    /// * If the proof input can't be built (first component of the pair is nil)
+    /// * If the proof input is not a proper list with length 6
+    fn get_cek_io_and_post_verify_function(
+        repl: &Repl<F>,
+        vars: Vec<Ptr>,
+        args: Vec<Ptr>,
+        body: Ptr,
+    ) -> Result<(Vec<Ptr>, Ptr)> {
+        let env = vars
+            .into_iter()
+            .zip(args)
+            .fold(repl.store.intern_nil(), |acc, (v, a)| {
+                repl.store.cons(repl.store.cons(v, a), acc)
+            });
+
+        let (io, ..) = repl
+            .eval_expr_with_env(body, env)
+            .with_context(|| "evaluating protocol body")?;
+
+        let Ptr::Tuple2(Tag::Expr(ExprTag::Cons), idx) = &io[0] else {
+            bail!(
+                "Protocol function must return a pair. Got {}",
+                io[0].fmt_to_string(&repl.store, &repl.state.borrow())
+            )
+        };
+        let (pre_verify, post_verify) = repl.store.fetch_2_ptrs(*idx).unwrap();
+
+        if pre_verify.is_nil() {
+            bail!("Pre-verification predicate rejected the input")
+        }
+
+        let Some((cek_io, None)) = repl.store.fetch_list(pre_verify) else {
+            bail!("Pre-verification predicate must return a proper list")
+        };
+        if cek_io.len() != 6 {
+            bail!("Pre-verification predicate return must have length 6")
+        }
+        Ok((cek_io, *post_verify))
+    }
+
+    /// Runs the post-verify predicate (if it's not nil)
+    ///
+    /// # Errors
+    /// * If the predicate is not a function
+    /// * If the predicate evaluation fails
+    /// * If the predicate rejects the proof (evaluation returns nil)
+    fn post_verify_check(repl: &Repl<F>, post_verify: &Ptr) -> Result<()> {
+        if !post_verify.is_nil() {
+            let Some((_, bod, env)) = repl.store.fetch_fun(post_verify) else {
+                bail!("Post-verification predicate must be a function")
+            };
+            // the following assumes Lurk Alpha encoding of function bodies
+            let (bod, _) = repl.store.car_cdr(bod)?;
+            let (io, ..) = repl
+                .eval_expr_with_env(bod, *env)
+                .with_context(|| "evaluating post-verify call")?;
+            if io[0].is_nil() {
+                bail!("Post-verification predicate rejected the input")
```

Comment:
Same comment as above. This shouldn't be an error. This should result in a successful rejection of the claim.